R/merge.R

```diff
@@ -2,37 +2,78 @@
 #'
-#' Is a secure way to merge the data and the meta data of two \code{GADSdat} objects. Currently, only limited merging options are supported.
+#' Is a secure way to merge the data and the meta data of two \code{GADSdat} objects.
+#' Currently, only limited merging options are supported.
 #'
 #' If there are duplicate variables (except the variables specified in the \code{by} argument), these variables are removed from y.
 #' The meta data is joined for the remaining variables via \code{rbind}.
 #'
+#' The function supports automatically recoding missing values created through merging with a designated missing code
+#' (\code{missingValue}) and a value label (\code{missingValLabel}).
+#'
 #'@param x \code{GADSdat} object imported via \code{eatGADS}.
 #'@param y \code{GADSdat} object imported via \code{eatGADS}.
 #'@param by A character vector.
 #'@param all A character vector (either a full join or an inner join).
 #'@param all.x See merge.
 #'@param all.y See merge.
+#'@param missingValue A numeric value that is used to replace missing values introduced through the merge.
+#'@param missingValLabel The value label that is assigned to all variables into which \code{missingValue} is inserted.
 #'@param ... Further arguments are currently not supported but have to be included for \code{R CMD} checks.
 #'
 #'@return Returns a \code{GADSdat} object.
 #'
 #'
 #'@export
-merge.GADSdat <- function(x, y, by, all = TRUE, all.x = all, all.y = all, ...) {
+merge.GADSdat <- function(x, y, by, all = TRUE, all.x = all, all.y = all,
+                          missingValue = NULL, missingValLabel = NULL, ...) {
   check_GADSdat(x)
   check_GADSdat(y)
-  if(!is.character(by)) stop(by, " is not a character vector.")
-  if(!all(by %in% names(x$dat))) stop(by, " is not a variable in x.")
-  if(!all(by %in% names(y$dat))) stop(by, " is not a variable in y.")
-  # drop double variables from y
+  if(!is.character(by)) {
+    stop(by, " is not a character vector.")
+  }
+  check_vars_in_GADSdat(x, vars = by, argName = "by", GADSdatName = "x")
+  check_vars_in_GADSdat(y, vars = by, argName = "by", GADSdatName = "y")
+
+  # drop duplicate variables from y
   y_vars <- c(by, names(y$dat)[!names(y$dat) %in% names(x$dat)])
-  if(!length(y_vars) > length(by)) stop("y does not contain unique variables.")
+  if(!length(y_vars) > length(by)) {
+    stop("y does not contain unique variables.")
+  }
 
-  newDat <- merge(x$dat, y$dat[, y_vars], by = by, all = all, all.x = all.x, all.y = all.y)
-  newLabels <- rbind(x$labels, y$labels[y$labels$varName %in% y_vars[!y_vars %in% by], ])
+  # replace NAs in original data sets to be able to recode them after merging
+  if(!is.null(missingValue)) {
+    check_numericArgument(missingValue, argName = "missingValue")
+
+    intermediate_code <- -99.9989
+    if(any(!is.na(x$dat) & x$dat == intermediate_code) || any(!is.na(y$dat) & y$dat == intermediate_code)) {
+      stop("Intermediate code is already in the data.")

```

Comment:
Hmmm... What would the user do in this case? 🤔 I guess it is highly improbable that the user would have such a code in their dataset. But maybe, there could be a solution offered, e.g., `"... Please replace entries of -99.9989 from your dataset"`? (Does this error message also require a test to be sure it fires somewhen?)

```r
 # with existing NAs without valLabel
  df1c <- df1
  df1c$dat[1, "V1"] <-  -99.9989
  df3c <- df3
  df3c$dat[2:3, "V2"] <- NA
  out3 <- merge(df1c, df3c, by = "ID1", missingValue = -99)
  expect_error(merge(df1c, df3c, by = "ID1", missingValue = -99), "Intermediate code is already in the data\\.")
```

Reply:
Alternatively, for this highly improbable case: Could it maybe be packed in some sort of loop that adds more digits (up to 10)? 🤔 (But this might be over-engineering!)

Reply:
Just asked AI and it also came up with a loop solution (from my opinion, it should not let the execution take longer as only a limited range of missings would be plausible):

```r
  existing_values <- c(x$dat, y$dat)
  intermediate_code <- -99.9989  # Default value
  while (any(!is.na(existing_values) & existing_values == intermediate_code)) {
    intermediate_code <- runif(1, -1e6, 1e6)  # Generate a random value in a wide range
  }
```

Reply:
I have implemented your solution as suggested. Thanks for taking the time and coming up with these alternatives!

Only tweaks I made to your suggestions:
- I pre-check for any `NAs` in the data now, to avoid performing any memory or time consuming computations 
- I avoid creating a vector `existing_values`, as for large data sets this could be memory intensive